File: backend/schemas/recommendations.py
from typing import Optional
from pydantic import BaseModel, Field
from core.config import settings


# ── Shared building blocks ─────────────────────────────────────────────────────

class UserAnswer(BaseModel):
    question: str
    answer: str


class DietaryContext(BaseModel):
    """
    Optional per-session dietary override.
    If the user is authenticated, their profile prefs are used automatically.
    This allows a guest user to still apply filters for a single session,
    or an authenticated user to temporarily override their profile.
    """
    dietary_flags: list[str] = Field(default_factory=list)
    allergens: list[str] = Field(default_factory=list)


# ── /generate-questions ────────────────────────────────────────────────────────

class GenerateQuestionsRequest(BaseModel):
    mood: str = Field(..., description=f"One of: {settings.VALID_MOODS}")


class QuestionOption(BaseModel):
    label: str
    value: str


class Question(BaseModel):
    question: str
    options: list[QuestionOption]


class GenerateQuestionsResponse(BaseModel):
    mood: str
    questions: list[Question]


# ── /get-recommendations ───────────────────────────────────────────────────────

class GetRecommendationsRequest(BaseModel):
    initial_mood: str = Field(..., description=f"One of: {settings.VALID_MOODS}")
    answers: list[UserAnswer]
    dietary_override: Optional[DietaryContext] = None  # Optional guest/override filters


class DishRecommendation(BaseModel):
    dish: str
    restaurant: str
    tags: list[str]
    reasoning: str
    foodpanda_link: str
    price_range: Optional[str] = None


class GetRecommendationsResponse(BaseModel):
    session_id: str
    refined_mood: str
    refined_tags: list[str]
    dietary_filters_applied: DietaryContext
    recommendations: list[DishRecommendation]

File: frontend/src/api/recommendations.js
import { apiRequest } from './client';

/**
 * Step 1: Get LLM-generated follow-up questions for a given mood.
 * @param {string} mood - one of VALID_MOODS
 * @returns {{ mood: string, questions: Array<{ question, options: [{label, value}] }> }}
 */
export async function generateQuestions(mood) {
  return apiRequest('/recommendations/generate-questions', {
    method: 'POST',
    body: { mood },
  });
}

/**
 * Step 2: Run the full recommendation pipeline.
 * Token is injected automatically by apiRequest if the user is logged in,
 * which causes the backend to load their dietary profile.
 *
 * @param {{
 *   initial_mood: string,
 *   answers: Array<{ question: string, answer: string }>,
 *   dietary_override?: { dietary_flags: string[], allergens: string[] }
 * }} payload
 * @returns {{
 *   session_id, refined_mood, refined_tags,
 *   dietary_filters_applied,
 *   recommendations: Array<{ dish, restaurant, tags, reasoning, foodpanda_link, price_range }>
 * }}
 */
export async function getRecommendations(payload) {
  return apiRequest('/recommendations/get-recommendations', {
    method: 'POST',
    body: payload,
  });
}

/**
 * Step 3: Submit thumbs up/down for a completed session.
 * @param {{ session_id: string, thumbs_up: boolean }} payload
 * @returns {{ success: boolean, message: string }}
 */
export async function submitFeedback(payload) {
  return apiRequest('/feedback/submit', {
    method: 'POST',
    body: payload,
  });
}